[OSPP] Alconna V2 - Sistana

GreyElaina opened the pull request:
resolve #88

GreyElaina committed:
sistana: migrate from local lab

GreyElaina commented:
敬请见证：https://wakatime.com/@9b922d20-ae54-4b73-81de-c31145268b46

GreyElaina committed:
sistana: new receiver, endpoint determine

GreyElaina committed:
sistana: feat: ConstraintRx

GreyElaina committed:
sistana: fix: push ahead with Value.value

GreyElaina committed:
sistana: free required snapshot from track

GreyElaina committed:
sistana: feat: complete_on_determined

GreyElaina committed:
sistana: feat: support option traverse

GreyElaina committed:
sistana: fix: apply default with Value.value

GreyElaina committed:
sistana: feat: spec
ial rx behavior

GreyElaina committed:
sistana: fix: always set value now

GreyElaina committed:
sistana: remove SubcommandTraverse.satisfied

GreyElaina committed:
sistana: feat: IndexedOptionTraversesRecord counts by keyword

GreyElaina committed:
sistana: feat: Buffer.pushleft

GreyElaina committed:
sistana: Buffer.pushleft reject empty segment

GreyElaina committed:
sistana: feat: option traverse works

GreyElaina committed:
sistana: analyzer returns description only

GreyElaina committed:
sistana: some tiny changes

GreyElaina committed:
sistana: fix: append -> appendleft

GreyElaina committed:
sistana: feat: Fragment.{separators, hybrid_separators}

GreyElaina committed:
sistana: clean done todo

GreyElaina committed:
sistana: feat: record subcommmand trigger in traverse

GreyElaina committed:
sistana: feat: Fragment.apply_nepattern(capture_mode: bool)

GreyElaina committed:
Merge branch 'ArcletProject:dev' into v2-sistana

GreyElaina committed:
legecy: merge arclet.alconna.model pyi

GreyElaina committed:
sistana: move type, cast attr to userspace

GreyElaina committed:
sistana: respect Fragment.cast

GreyElaina committed:
sistana: complete `__init__` modules

GreyElaina committed:
sistana: new snapshot gen

GreyElaina committed:
sistana: SubcommandPattern.{compacts => compact_keywords}

GreyElaina committed:
sistana: feat: SubcommandPattern.build

RF-Tar-Railt commented on src/arclet/alconna/sistana/analyzer.py:
```python
    def loopflow(self, snapshot: AnalyzeSnapshot[T], buffer: Buffer[T]) -> LoopflowDescription:
        while True:
            traverse = snapshot.traverses[-1]
            context = traverse.subcommand
            mix = traverse.mix

            if snapshot.determined and self.complete_on_determined and mix.satisfied:
                return LoopflowDescription.completed

            pointer_type, pointer_val = traverse.ref.data[-1]

            try:
                token = buffer.next(traverse.subcommand.separators)
            except OutOfData:
                if mix.satisfied:
                    mix.complete_all()

                    # 在 option context 里面，因为 satisfied 了，所以可以直接返回 completed。
                    # 并且还得确保 option 也被记录于 activated_options 里面。
                    if pointer_type == "option":
                        mix.tracks[pointer_val].complete()
                        option_traverse = traverse.option_traverses[-1]
                        option_traverse.completed = True
                        traverse.ref = traverse.ref.parent

                        if context.options[pointer_val].allow_duplicate:
                            mix.pop_track(pointer_val)

                    snapshot.determine(traverse.ref)
                    return LoopflowDescription.completed

                # 这里如果没有 satisfied，如果是 option 的 track，则需要 reset
                # 从 Buffer 吃掉的东西？我才不还。
                if pointer_type == "option":
                    mix.reset(pointer_val)

                return LoopflowDescription.unsatisfied

            if pointer_type == "prefix":
                if not isinstance(token.val, str):
                    return LoopflowDescription.header_expect_str

                if context.prefixes is not None:
                    prefix = context.prefixes.get_closest_prefix(buffer.runes[0])  # type: ignore
                    if prefix == "":
                        return LoopflowDescription.prefix_mismatch

                    token.apply()
                    buffer.pushleft(token.val[len(prefix) :])

                traverse.ref = traverse.ref.parent.header()  # 直接进 header.
            elif pointer_type == "header":
                if not isinstance(token.val, str):
                    return LoopflowDescription.header_expect_str

                token.apply()

                if token.val == context.header:
                    pass  # do nothing
                elif context.compact_header and token.val.startswith(context.header):
                    # ahead 似乎并不能很好的处理这种情况：这里目的不是为了分「完全已知」的段落，而是分一半 —— 去掉并留置。
                    # 再次重申，ahead 里的所有段落全都是「已经分好」了的。
                    # 在 perfix 和 header 中，我们需要对 buffer 的第一段进行直接替换，而 ahead 倾向于 **已经** 分好段的情况。
                    # ~~而除了 compact subcommand / option 外，对于 header 和 prefix，可以选择直接操作 runes[0]。~~

                    v = token.val[len(context.header) :]
                    if v:
                        buffer.runes.insert(0, v)
                else:
                    return LoopflowDescription.header_mismatch

                traverse.ref = traverse.ref.parent
            elif isinstance(token.val, str):
                if pointer_type == "subcommand":
                    if token.val in context.subcommands:
                        subcommand = context.subcommands[token.val]

                        if mix.satisfied or not subcommand.satisfy_previous:
                            token.apply()
                            mix.complete_all()

                            # context hard switch
                            snapshot.traverses.append(
                                SubcommandTraverse(
                                    subcommand,
                                    token.val,
                                    traverse.ref.subcommand(subcommand.header),
                                    subcommand.preset.new_mix(),
                                )
                            )
                            continue
                        elif not subcommand.soft_keyword:
                            return LoopflowDescription.unsatisfied_switch_subcommand
                        # else: soft keycmd，直接进 mainline
                    elif token.val in context.options:
                        option = context.options[token.val]

                        # 之前的版本中如果 track 没有 satisfied，就会继续进入 track forward 流程。
                        # 这里有个问题，如果 Fragments 里有个 variadic，那么就会一直进入 forward 流程 —— 这就不太对了，所以我删掉了。
                        if not option.soft_keyword or mix.satisfied:
                            if context.preset.tracks[option.keyword]:
                                # 仅当需要解析 fragments 时进行状态流转。
                                traverse.ref = traverse.ref.option(token.val)

                            if not option.allow_duplicate and option.keyword in traverse.option_traverses:
                                return LoopflowDescription.option_duplicated

                            traverse.option_traverses.append(
                                OptionTraverse(
                                    trigger=token.val,
                                    is_compact=False,
                                    completed=not context.preset.tracks[option.keyword],
                                    option=option,
                                    track=mix.tracks[option.keyword],
                                )
                            )

                            token.apply()  # 在最后才 apply，因为上面会根据 duplicate 的情况判定 panic，不能一开始就吃掉。
                            continue
                        # else: What?
                    # else: What?
                else:  # pointer_type == "option"
                    option_traverse = traverse.option_traverses[-1]

                    if token.val in context.subcommands:
                        # 当且仅当 option 已经 satisfied 时才能让状态流转进 subcommand。
                        # subcommand.satisfy_previous 处理起来比较复杂，这里先 reject。

                        # 对于 OptionPattern.allow_duplicate。
                        subcommand = context.subcommands[token.val]
                        option = context.options[pointer_val]
                        track = mix.tracks[option.keyword]
                        
                        if track.satisfied:
                            track.complete()
                            traverse.ref = traverse.ref.parent
                            option_traverse.completed = True

                            if option.allow_duplicate:
                                mix.pop_track(option.keyword)

                            if mix.satisfied:
                                token.apply()
                                mix.complete_all()

                                # context hard switch
                                snapshot.traverses.append(
                                    SubcommandTraverse(
                                        subcommand,
                                        token.val,
                                        traverse.ref.subcommand(token.val),
                                        subcommand.preset.new_mix(),
                                    )
                                )
                                continue
                            elif not subcommand.soft_keyword:
                                return LoopflowDescription.unsatisfied_switch_subcommand
                        elif not subcommand.soft_keyword:
                            mix.reset(option.keyword)
                            return LoopflowDescription.switch_unsatisfied_option
                        # else: soft keycmd，直接进 mainline ?

                    elif token.val in context.options:
                        option = context.options[token.val]
                        track = mix.tracks[option.keyword]

                        if not track.satisfied:
                            if not option.soft_keyword:
                                mix.reset(option.keyword)
                                return LoopflowDescription.previous_unsatisfied
                            # else: soft keycmd，直接进 mainline ?
                        else:
                            track.complete()
                            traverse.ref = traverse.ref.parent
                            option_traverse.completed = True

                            if option.allow_duplicate:
                                mix.pop_track(option.keyword)

                            continue

                    if context.compacts is not None:
                        prefix = context.compacts.get_closest_prefix(token.val)
                        if prefix:
                            # 这里仍然需要关注 soft_keycmd 和 satisfied 的情况。
                            # 这里有个有趣的点……至少三方因素会参与到这里，所以逻辑关系会稍微复杂那么一点。
                            # 我加了一个 Track.assignable，这样我们就能知道是否还有 fragments 可供分配了。

                            # 我想了想，soft keyword 不会影响这个 —— token.val 根本不是关键字（如果是就不会在这个分支了）。
                            redirect = False

                            if prefix in context.subcommands:
                                # 老样子，需要 satisfied 才能进 subcommand，不然就进 track forward 流程。
                                redirect = mix.satisfied
                            elif prefix in context.options:
                                # NOTE: 这里其实有个有趣的点需要提及：pattern 中的 subcommands, options 和这里的 compacts 都是多对一的关系，
                                # 所以如果要取 track 之类的，就需要先绕个路，因为数据结构上的主索引总是采用的 node 上的单个 keyword。
                                option = context.options[prefix]
                                track = mix.tracks[option.keyword]

                                redirect = track.assignable or option.allow_duplicate
                                # 这也排除了没有 fragments 设定的情况，因为这里 token.val 是形如 "-xxx11112222"，已经传了一个 fragment 进去。

                            # else: 你是不是手动构造了 TrieHard？
                            # 由于默认 redirect 是 False，所以这里不会准许跳转。

                            if redirect:
                                token.apply()
                                prefix_len = len(prefix)
                                buffer.ahead.appendleft(token.val[:prefix_len])
                                buffer.pushleft(token.val[prefix_len:])
                                continue
                        # else: What?
                    # else: What?
            elif pointer_type == "subcommand":
                track = mix.tracks[context.header]

                try:
                    response = track.forward(buffer, context.separators)
                except OutOfData:
                    # 称不上是 context switch，continuation 不改变 context。
                    return LoopflowDescription.out_of_data_subcommand
                except Rejected:
                    raise
                except ParsePanic:
                    raise
                except Exception as e:
                    raise ParsePanic from e
                else:
                    if response is None:
                        # track 上没有 fragments 可供分配了，此时又没有再流转到其他 traverse
                        return LoopflowDescription.unexpected_segment
                    # else: What?
            else:  # pointer_type == "option"
                # option fragments 的处理是原子性的，整段成功才会 apply changes，否则会被 reset。
                option = context.options[pointer_val]
                track = mix.tracks[option.keyword]

                if traverse.option_traverses.count(option.keyword) > 1:
                    rx_getter = traverse.option_traverses[-2].track._assign_getter(track.fragments[0].name)
                else:
                    rx_getter = None

                try:
                    response = track.forward(buffer, option.separators, rx_getter)
                except OutOfData:
                    mix.reset(option.keyword)
                    return LoopflowDescription.out_of_data_option
                except Rejected:
                    raise
                except ParsePanic:
                    raise
                except Exception as e:
                    raise ParsePanic from e
                else:
                    if response is None:
                        # track 上没有 fragments 可供分配了。
                        # 这里没必要 complete：track.complete 只是补全 assignes。

                        traverse.ref = traverse.ref.parent
                        traverse.option_traverses[-1].completed = True

                        if option.allow_duplicate:
                            mix.pop_track(option.keyword)
                        # else: What?
                    # else: What?
            # next loop
```

RF-Tar-Railt replied:
以及按目前代码我感觉你都不用写一个class Analyser了，直接把loopflow 拿出来（

GreyElaina replied:
你的 else branch question 的答案全是进 track  process 或者是进下一个循环。

RF-Tar-Railt commented on src/arclet/alconna/sistana/utils/misc.py:
```diff
@@ -0,0 +1,43 @@
+E = TypeVar("E", bound=BaseException)
+
+
+@contextmanager
+def cvar(var: ContextVar[T], val: T):
+    token = var.set(val)
+    try:
+        yield val
+    finally:
+        var.reset(token)
```
tarina 里有个的
https://github.com/ArcletProject/Tarina/blob/c71cae3a8a5e4751decfa48aab19939d0091f97f/src/tarina/context.py#L11

GreyElaina replied:
这个没用 —— 并且 BCC 也有一个和你 Tarina 这个完全一致的版本。

并且目前就我个人观点，我认为写成这个 `cvar` 反而是更简洁许多的。

只是这个并没有在 sistana 中用到，后续我会把这个删掉。

RF-Tar-Railt commented on src/arclet/alconna/sistana/fragment.py:
```diff
@@ -0,0 +1,72 @@
+
+        pat = type_parser(self.type.value)
+
+        if capture_mode:
+            self.capture = RegexCapture(pat.alias)
```
if pat.mode in (MatchMode.REGEX_MATCH, MatchMode.REGEX_CONVERT):
    self.capture = RegexCapture(pat.regex_pattern)
else:
    self.capture = RegexCapture(pat.alias)  # ?

RF-Tar-Railt reviewed (commented):
model这边加点comment或者docstring

GreyElaina committed:
sistana: remove unused util

GreyElaina committed:
chore: reimpl safe_dcls_kw

GreyElaina committed:
sistana: mis-operate git

GreyElaina committed:
sistana: capture nepattern with pat.regex_pattern

GreyElaina committed:
sistana: add comments to CLARIFY

GreyElaina committed:
sistana: fix {option -> option'}; stop obtain aliases for pointer

GreyElaina committed:
sistana: export Value type

GreyElaina committed:
fix: ensure prefix: str to satisfy sistana

GreyElaina reviewed (commented):
> 以及按目前代码我感觉你都不用写一个class Analyser了，直接把loopflow 拿出来（ 

关于这个我有异议：我计划为 Analyzer 提供一些 option 以控制其行为，虽然现 Flywheel 在处理 descriptor 上仍有欠缺，在 Analyzer 上设置 Flywheel Anycast 或许是不错的想法。

GreyElaina committed:
sistana: LoopflowDescription.__repr__

GreyElaina committed:
stargazing: basically works

GreyElaina committed:
stargazing: remove unused expression

GreyElaina committed:
sistana: remove duplicated util

GreyElaina committed:
sistana: fix nepattern apply
stargazing: fix nepattern apply

GreyElaina committed:
sistana: drop getter/putter generator in track

GreyElaina committed:
sistana: make pyright happy

GreyElaina committed:
sistana-compat: feat: soft_keyword for legecy

GreyElaina committed:
chore: ruff line length = 140

GreyElaina committed:
sistana: fix: import nepattern

GreyElaina committed:
sistana: pointer role enum

GreyElaina committed:
sistana: move some continuable description

GreyElaina committed:
sistana: remove commented-out Mix.result

GreyElaina committed:
sistana: rename param for Track.fetch

GreyElaina committed:
sistana: rename Track.forward({rxget, rxput})

GreyElaina committed:
sistana: fix wrong comment

GreyElaina committed:
Merge branch 'ArcletProject:dev' into v2-sistana

GreyElaina committed:
docs: changelog

Add Sistana command parsing engine

This commit adds the Sistana command parsing engine, which is a redesigned next-generation command parsing engine. It supports soft keywords, snapshots, and other new features, providing more efficient prefix matching, compact parameters/instructions, and analysis interruption capabilities.

- This feature is currently experimental and not enabled by default.
- Sistana may conflict with or disable certain Alconna features.
- Sistana has not been highly optimized or utilized its features, so there may be performance issues.
- Sistana (and its upstream libraries) only support Python 3.9 and above.
- Compatibility with various features of Sistana has not been fully tested.

GreyElaina committed:
sistana: use elaina-segment buffer (cython)

GreyElaina committed:
sistana: enable dataclass slots=True or hand-written

GreyElaina committed:
sistana: perf: hand-written init and slots

GreyElaina committed:
chore: remove unused imports

GreyElaina committed:
sistana: merge error catch

GreyElaina committed:
sistana: new receiver (fetch, prev, put)

GreyElaina committed:
sistana: fix: access first rune

GreyElaina committed:
sistana: cmp pointer type with `is`

GreyElaina committed:
sistana: rename module {track => mix}

GreyElaina committed:
sistana: feat: option-level receiver & mix support

GreyElaina committed:
sistana: fix: net assignes with frag.name

GreyElaina committed:
sistana: move/simplify util module

GreyElaina committed:
sistana: remove net

GreyElaina committed:
sistana: preset accepts Track only

GreyElaina committed:
sistana: feat: Track.header_fragment, Option.{header_fragment, keep_previous_assignes}

GreyElaina committed:
sistana: fix pointer repr

GreyElaina committed:
sistana: ok we have var pos for fragments

GreyElaina committed:
sistana: re-export from upstream

GreyElaina committed:
sistana: apply posvar

GreyElaina committed:
sistana: allow compact aliases

GreyElaina committed:
sistana: header fragment can fetch trigger segment

GreyElaina committed:
sistana: feat: subcommand header fragment

GreyElaina committed:
sistana: fix: SubcommandPattern.build didn't init empty track for empty fragments

GreyElaina committed:
sistana: feat: option header separators

GreyElaina committed:
sistana: add fragments check for header_separators

GreyElaina committed:
sistana: always create track for option

GreyElaina committed:
sistana: format code

GreyElaina committed:
sistana: fix: check fragments order when add_track

GreyElaina committed:
sistana: remove unused reject helper

GreyElaina committed:
sistana: fix: Pointer should impl eq and hash

GreyElaina committed:
sistana: feat: export

GreyElaina committed:
sistana-stargazing: feat: loopflow exit reason to legecy alconna exceptions

GreyElaina committed:
sistana: rename LoopflowDescription => LoopflowExitReason

GreyElaina committed:
sistana: oops i typo const

GreyElaina committed:
sistana: feat: emit_header respect allow_deplicate

GreyElaina committed:
sistana: widen prefix logic

GreyElaina committed:
sistana: feat: Pointer.{last_role, last_value}

GreyElaina committed:
sistana-stargazing: alconna patch minimal works

GreyElaina committed:
sistana-stargazing: buggy patch

GreyElaina commented:
别急，转 Ready for Review / Merge 前我先看看能修多少个测试

GreyElaina committed:
sistana: patch with debug method controlled by param

GreyElaina commented:
将进行一次大的重构，添加新的功能与更健全的状态流转设计

GreyElaina committed:
refactor(sistana): new snapshot model

fixme: poor performance
todo: _compact_keywords
todo: separator_optbind
todo: sort some non-pattern attrs

GreyElaina committed:
sistana: better main/alter processing

GreyElaina committed:
sistana: improve prof

GreyElaina committed:
sistana: improve perf

GreyElaina committed:
feat(sistana): snapshot enter_{subcommand, option}

GreyElaina committed:
sistana: improve prof

GreyElaina committed:
sistana: improve performance

GreyElaina committed:
sistana: improve performance, 1-10% less

GreyElaina committed:
sistana: remove unused

GreyElaina committed:
sistana: improve perf, remove unused / unnecessary

GreyElaina committed:
feat(sistana): implement option separator

RF-Tar-Railt commented on src/arclet/alconna/model.py:
```diff
@@ -1,34 +1,80 @@
 
 
 @dataclass(init=False, eq=True)
 class Sentence:
+    """句段
+
+    句段由 `Analyser` 编译而来, 代表选项或者子命令的需求前缀。
+
+    Attributes:
+        name (str): 句段名称
+    """
+
     __slots__ = ("name",)
     __str__ = lambda self: self.name  # type: ignore
     __repr__ = lambda self: self.name  # type: ignore
 
-    def __init__(self, name):
+    if TYPE_CHECKING:
+        name: str
+
+    def __init__(self, name: str) -> None:
         self.name = name
```
删了吧）

GreyElaina replied:
对 arclet.alconna.[^(sistana)(_stargazing)] 的修改都是参考用的……

RF-Tar-Railt commented on src/arclet/alconna/sistana/err.py:
```diff
@@ -0,0 +1,46 @@
+class TransformPanic(ParsePanic): ...
+
+
+class ReceivePanic(ParsePanic): ...
```
如果可以的话，写到alconna.exception里

GreyElaina committed:
sistana: simplify logic, swap branches and improve perf

GreyElaina committed:
sistana: drop triehard, implement compact

GreyElaina committed:
sistana: use cached trigger

GreyElaina committed:
sistana: remove unused import

GreyElaina committed:
sistana: fix option trigger

GreyElaina committed:
test: sistana

GreyElaina committed:
sistana: fix: fragments order check

GreyElaina committed:
sistana: fix non-consist naming

GreyElaina committed:
sistana: fix wrong exit reason

GreyElaina committed:
sistana: fix determined instant exit

GreyElaina committed:
sistana: add tests

GreyElaina committed:
sistana: remove unused args

GreyElaina committed:
sistana: remove unused protected helper method

GreyElaina committed:
sistana: fix satisfied check

GreyElaina committed:
sistana: fix header edge case

GreyElaina committed:
Refactor test_analyze.py and test_pattern.py

- Remove unused protected helper method in test_analyze.py
- Remove unused args in test_analyze.py
- Add tests for header separator and stage unsatisfied in test_analyze.py
- Add tests for header edge case and variadic in test_mix.py
- Add tests for aliases and add_option in test_pattern.py
- Add tests for AccumRx, CountRx, and ConstRx in test_receiver.py

GreyElaina committed:
sistana: basically acceptable

GreyElaina committed:
sistana: format

GreyElaina committed:
test: ruffed

GreyElaina committed:
Update dependencies in pyproject.toml

GreyElaina committed:
Update Python versions in test workflow

RF-Tar-Railt commented on pyproject.toml:
```diff
@@ -139,3 +142,6 @@ pythonVersion = "3.8"
 pythonPlatform = "All"
```
这里改成 3.9

RF-Tar-Railt commented on src/arclet/alconna/sistana/fragment.py:
```diff
@@ -0,0 +1,97 @@
+
+        from nepattern import MatchMode
+
+        if capture_mode:
+            if pat.mode in (MatchMode.REGEX_MATCH, MatchMode.REGEX_CONVERT):
+                self.capture = RegexCapture(pat.regex_pattern)
+            else:
+                self.capture = RegexCapture(pat.alias)  # type: ignore
+        else:
+
+            def _validate(v: Segment):
+                if isinstance(v, (Quoted, UnmatchedQuoted)):
+                    if isinstance(v.ref, str):
+                        v = str(v)
+                    else:
+                        v = v.ref
+
+                return pat.validate(v).success
+
+            self.validator = _validate
+
+        if self.cast:
+
+            def _transform(v: Segment):
+                return pat.validate(str(v)).value()
+
+            self.transformer = _transform
```
https://github.com/ArcletProject/Alconna/commit/3e1a49e038f1fc0c9c1bc86b2114f313b1fbb07e#diff-aad2c1776c92e0bc33c357affa6830f52b94c1c25da92a2b18be2ebbf217cbe8L6-R98

GreyElaina committed:
chore: bump pyright target python to 3.9

GreyElaina committed:
sistana: fix nepattern
